src/gui/macOS/fileprovider.h

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (C) by Claudio Cambra <claudio.cambra@nextcloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#pragma once
+
+#include <QObject>
+
+#include "fileproviderdomainmanager.h"
+#include "fileprovidersocketserver.h"
+
+namespace OCC {
+
+class Application;
+
+namespace Mac {
+
+// NOTE: For the file provider extension to work, the app bundle will
+// need to be correctly codesigned!
+
+class FileProvider : public QObject
+{
+    Q_OBJECT
+
+public:
+    static FileProvider *instance();
+    ~FileProvider() override;
+
+    static bool fileProviderAvailable();
+
+private:
+    std::unique_ptr<FileProviderDomainManager> _domainManager;
+    std::unique_ptr<FileProviderSocketServer> _socketServer;
```

Comment:
Because there are situations where we have this class included in the final application but we do not want to create the objects

Say we detect we are running on macOS 10.15 which does not support File Provider, we do not want to bother with creating the domain manager or the socket server. Using the pointers we can just leave them as null and not have to forcibly create instances for nothing